bin/varnishd/cache/cache_req_body.c

```diff
@@ -43,15 +43,63 @@
 #include "storage/storage.h"
 #include "hash/hash_slinger.h"
 
+/*----------------------------------------------------------------------
+ * Check and potentially update req framing headers.
+ */
+
+static ssize_t
+vrb_cached(struct req *req, ssize_t req_bodybytes)
+{
+	ssize_t l0, l;
+
+	CHECK_OBJ_NOTNULL(req, REQ_MAGIC);
+	AZ(req->req_body_cached);
+	assert(req_bodybytes >= 0);
+
+	l0 = http_GetContentLength(req->http0);
+	l = http_GetContentLength(req->http);
+
+	assert(req->req_body_status->avail > 0);
+	if (req->req_body_status->length_known) {
+		if (req->req_body_partial) {
+			assert(req_bodybytes < l0);
+			assert(req_bodybytes < l);
+		} else {
+			assert(req_bodybytes == l0);
+			assert(req_bodybytes == l);
+		}
+		AZ(http_GetHdr(req->http0, H_Transfer_Encoding, NULL));
+		AZ(http_GetHdr(req->http, H_Transfer_Encoding, NULL));
+	} else if (!req->req_body_partial) {
+		/* We must update also the "pristine" req.* copy */
+		AZ(http_GetHdr(req->http0, H_Content_Length, NULL));
+		assert(l0 < 0);
+		http_Unset(req->http0, H_Transfer_Encoding);
+		http_PrintfHeader(req->http0, "Content-Length: %ju",
+		    (uintmax_t)req_bodybytes);
+
+		AZ(http_GetHdr(req->http, H_Content_Length, NULL));
```

Comment:
Also FTR:
The `req.http0` changes are logged, which confused me at first thinking that the `vrb_pull()` might be called twice:
```
**** v1    vsl|       1004 ReqUnset        c Transfer-encoding: chunked
**** v1    vsl|       1004 ReqHeader       c Content-Length: 110
**** v1    vsl|       1004 ReqUnset        c Transfer-encoding: chunked
**** v1    vsl|       1004 ReqHeader       c Content-Length: 110
```